server/src/main/java/org/elasticsearch/index/query/ReverseNestedQueryBuilder.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.query;
+
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.join.BitSetProducer;
+import org.apache.lucene.search.join.ToChildBlockJoinQuery;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.TransportVersion;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.lucene.search.Queries;
+import org.elasticsearch.index.mapper.NestedObjectMapper;
+import org.elasticsearch.index.search.NestedHelper;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.TransportVersions.NESTED_KNN_VECTOR_QUERY_V;
+import static org.elasticsearch.search.SearchService.ALLOW_EXPENSIVE_QUERIES;
+
+public class ReverseNestedQueryBuilder extends AbstractQueryBuilder<ReverseNestedQueryBuilder> {
```

Comment:
I can't see it used anywhere, is it? Or is it intended to be used by users? I can understand why aggregations need a reverse nested aggregation to be able to bucket on child values and aggregate on parent values or vice-versa, but I'm not sure I get the use-case for a reverse nested query?

Reply:
@jpountz I may remove it for now. The idea is being able to have flexible prefiltering with knn.

Reply:
Oh I see, so that users would be able to filter on parent documents from the nested filter? E.g. assuming this gets exposed API-wise in a similar way as `sort`, you could do something like this?

```javascript
POST _search
{
  "knn": {
    "field": "comments.vector" // comments is nested
    "nested": {
      "path": "comments", // the path to the nested object
      "filter": { // pre-filter on nested objects
        "bool": {
          "filter": [
            {
              "match": {
                "comments.author": "benwtrent"
              }
            },
            {
              "reverse_nested": {
                "match": {
                  "is": "open"
                }
              }
            }
          ]
        }
      }
    }
  }
}
```

Thinking out loud: Or maybe we could still allow using the `filter` element when nested gets used and automatically combine the nested filter with the top-level filter?

```javascript
POST _search
{
  "knn": {
    "field": "comments.vector" // comments is nested
    "nested": {
      "path": "comments", // the path to the nested object
      "filter": { // pre-filter on nested objects
        "match": {
          "comments.author": "benwtrent"
        }
      }
    },
    "filter": { // gets combined with the nested filter via a ToChildBlockJoinQuery internally
      "match": {
        "is": "open"
      }
    }
  }
}
```

(This assumes a GitHub dataset of issues where docs look like below:

```javascript
{
  "title": "nested support for knn"
  "is": [ "open", "pr" ],
  "comments": [
    ...
    {
      "vector": [ ... ],
      "author": "benwtrent"
    }
    ...
  ]
}
```
)

Reply:
@jpountz I don't mind that nested option at all. It honestly is way easier than wrangling `reverse_nested`... blah blah. A very common use case here is filtering by top level.

I think though if we did the `nested` option, I don't think we need the ` "path": "comments", // the path to the nested object` part, we know that by the vector field path already no?

Reply:
Another option is that "filter" is always outside the nested context and to do a nested filter, it must have a "nested" query in it.

That doesn't seem too bad to me.

Reply:
OK, I poked around and I cannot figure out a way to support a top level filter without `reverse_nested`.

One option would be trying to figure out how to apply the `filter` parameter on the "parentsFilter". But this seems brittle and would still require the requisite children to be filtered somehow. Combining both results in an "empty set" as they are orthogonal document sets :/

Adding two new top-level options to knn to simply support nested seems very frustratingly expensive to me (we already have added "inner_hits").

@jimczi do you have any ideas here? Would like to try and get this over the line for 8.11. 

`reverse_nested` is complicated, I agree. We will have to figure out simplifications in the future unless y'all have a better idea that I am missing :(

Reply:
> I think though if we did the nested option, I don't think we need the  "path": "comments", // the path to the nested object part, we know that by the vector field path already no?

I agree we can infer it when there is a single level of nesting, which should be the most common case.

FWIW I don't think `reverse_nested` is too complicated, I was more looking for consitency with e.g. nested sorts which also need the ability to select some particular nested objects and did not require a `reverse_nested` query.

Thinking out loud: in general, users configure `nested` objects because they want to be able to filter at the nested level, something you can't do with regular objects. But here it's different, we're mostly using `nested` because of the limitation that inference can work on a single passage at once, and as a way to retrieve which passage matched via inner hits. So what about skipping filtering at the nested level for now and looking into it later?

Reply:
>  because of the limitation that inference can work on a single passage at once, and as a way to retrieve which passage matched via inner hits. So what about skipping filtering at the nested level for now and looking into it later?

@jpountz Your suggestion would be internally to only use `ToChildBlockJoinQuery` to wrap the passed filter and pass that to the child filter correct?

I don't think there is any other way to do it, but that could work.

Reply:
Correct, you would take the `filter` under the `knn` object which matches parent documents, then wrap it into a `ToChildBlockJoinQuery` so that it would match all child document of parent documents that match the filter, and pass this wrapper filter as the filter of a Lucene `KnnFloatVectorQuery` to find matching vectors.

Reply:
understood. This makes sense to me for now.

Reply:
If we're expecting the field to be declared in a nested context, I am in favour of not trying to add it when we perform the query. `knn` is a top level section so we need to magically wrap it with `nested`. If `knn` was a query, we'd expect the use of  `nested` explicitly. So inside a `knn` section I'd expect that we are in the `nested` context if the underlying knn field is `nested`. If we start the other way around then adding the ability to filter at the children level (nested) would not be possible without breaking. 
`reverse_nested` is not ideal but very powerful so I'd say that we can try to simplify later but I don't see why it couldn't be exposed in this first iteration? Maybe we need a better name though like `parent`?

Reply:
> So inside a knn section I'd expect that we are in the nested context if the underlying knn field is nested

FWIW this is not intuitive to me, and I suspect that users would get confused that they don't directly have access to their top-level fields in the `filter` element if the field that they are querying is nested.

This would also make it more challenging to hide the fact that we are using `nested` fields only as an implementation detail to store multiple vectors per document as users would need to adjust their queries.

If we want to add support for filtering at the nested level, I would make it explicit, like we do for queries, ie. something like this for queries

```javascript
"bool": {
  "must": {
    "nested": {
      "path": "comments"
      "query": < filter at the nested level >
    }
  },
  "filter": <filter at the parent level>
}
```

would translate to something like that for NN search:

```javascript
"knn": {
  "nested": {
    "path": "comments"
    "filter": < filter at the nested level >
  },
  "filter": <filter at the parent level>
}
```

Reply:
I see so what you propose is that we keep `filter` to always work at the root level and we leave the option open to add a `nested` section to filter at that level as a follow up? 

> This would also make it more challenging to hide the fact that we are using nested fields only as an implementation detail to store multiple vectors per document as users would need to adjust their queries

My point was that we shouldn't try to hide it. If it's a nested field it should work like a nested field. 
If we want to take the road of hiding it then we should not require to declare the field as nested?

Reply:
> what you propose is that we keep filter to always work at the root level and we leave the option open to add a nested section to filter at that level as a follow up?

This is correct.

> My point was that we shouldn't try to hide it. If it's a nested field it should work like a nested field.
If we want to take the road of hiding it then we should not require to declare the field as nested?

Agreed. I was referring to the idea of having a semantic text field that would automatically index passages as nested docs under the hood. So users would indeed not declare the field as nested in that case.

Reply:
Yes it does thanks, I am still not used to having `knn` as a top level section ;)